### PR TITLE
[cominterop] Handle NULL pointers when marshalling native-to-managed return values.

### PR DESCRIPTION
**Pull request by zfigura**
Fixes a bug running Rak24u with wine-mono: <https://bugs.winehq.org/show_bug.cgi?id=47561>

### CODE REVIEW

File: mono/metadata/cominterop.c [hunk 1]
```diff
@@ -2359,12 +2359,24 @@ cominterop_get_managed_wrapper_adjusted (MonoMethod *method)
...
+			MonoClass *rclass;
+
+			mono_mb_emit_stloc (mb, retval);
+			mono_mb_emit_ldarg (mb, sig_native->param_count-1);
+			pos_null = mono_mb_emit_branch (mb, CEE_BRFALSE);
+
+			mono_mb_emit_ldarg (mb, sig_native->param_count-1);
+			mono_mb_emit_ldloc (mb, retval);
+
+			rclass = mono_class_from_mono_type_internal (sig->ret);
```
Comment (jaykrell) [inline mono/metadata/cominterop.c]: Is there reason this separate from local variable?

File: mono/metadata/cominterop.c [hunk 2]
```diff
@@ -2359,12 +2359,24 @@ cominterop_get_managed_wrapper_adjusted (MonoMethod *method)
...
 		if (!preserve_sig) {
-			MonoClass *rclass = mono_class_from_mono_type_internal (sig->ret);
+			int pos_null;
+			MonoClass *rclass;
+
+			mono_mb_emit_stloc (mb, retval);
+			mono_mb_emit_ldarg (mb, sig_native->param_count-1);
+			pos_null = mono_mb_emit_branch (mb, CEE_BRFALSE);
+
+			mono_mb_emit_ldarg (mb, sig_native->param_count-1);
```
Comment (jaykrell) [inline mono/metadata/cominterop.c]: Spaces around binary operators please -- `param_count - 1` and above.

File: mono/metadata/cominterop.c [hunk 3]
```diff
@@ -2335,18 +2335,18 @@ cominterop_get_managed_wrapper_adjusted (MonoMethod *method)
 
 #ifndef DISABLE_JIT
 	if (!preserve_sig)
+	{
```
Comment (jaykrell) [inline mono/metadata/cominterop.c]: I like this brace placement, but around here it is just me.

File: mono/metadata/cominterop.c (hunk 1 above)
Comment (zfigura) [reply mono/metadata/cominterop.c]: Do you mean the C89-style declaration? No particular reason, I guess; it seems Mono usually uses C89-style declarations, except when it doesn't, and I'm personally more used to C89.

Comment (jaykrell) [reply mono/metadata/cominterop.c]: Yes. Mono is very C99. Uninitialized locals are bug farms imho.
Though I agree C99 is hardly worth anything over C89.
I'd prefer C++ even C++96.

File: mono/metadata/cominterop.c (hunk 2 above)
Comment (zfigura) [reply mono/metadata/cominterop.c]: Yep, that was copied from the previous code. Fixed now.

File: mono/metadata/cominterop.c (hunk 3 above)
Comment (zfigura) [reply mono/metadata/cominterop.c]: Fixed.

File: mono/metadata/cominterop.c (hunk 1 above)
Comment (zfigura) [reply mono/metadata/cominterop.c]: Okay, fixed.

File: mono/metadata/cominterop.c [hunk 4]
```diff
@@ -2359,12 +2358,23 @@ cominterop_get_managed_wrapper_adjusted (MonoMethod *method)
 
 	if (!MONO_TYPE_IS_VOID (sig->ret)) {
 		if (!preserve_sig) {
+			int pos_null;
+
+			mono_mb_emit_stloc (mb, retval);
+			mono_mb_emit_ldarg (mb, sig_native->param_count - 1);
+			pos_null = mono_mb_emit_branch (mb, CEE_BRFALSE);
```
Comment (jaykrell) [inline mono/metadata/cominterop.c]: I prefer `const int pos_null = …` but ok either way.

Comment (zfigura) [reply mono/metadata/cominterop.c]: Sure, done.

File: mono/tests/libtest.c [hunk 5]
```diff
@@ -3655,6 +3662,29 @@ mono_test_marshal_array_ccw_itest (int count, MonoComObject ** ppUnk)
...
+	if (!pUnk)
+		return 1;
+
+	hr = pUnk->vtbl->IntOut (pUnk, &i);
+	if (hr != 0)
+		return 2;
+	if (i != 33)
+		return 3;
+	if (test_null)
+	{
```
Comment (jaykrell) [inline mono/tests/libtest.c]: Another brace but ok.

File: mono/tests/libtest.c [hunk 6]
```diff
@@ -3512,6 +3513,11 @@ Return22NoICall(MonoComObject* pUnk)
 	return 22;
 }
 
+LIBTEST_API int STDCALL
+IntOut(MonoComObject* pUnk, int *a)
```
Comment (jaykrell) [inline mono/tests/libtest.c]: Usually mono has space after paren, but this file does not, granted.

File: mono/tests/libtest.c (hunk 5 above)
Comment (jaykrell) [reply mono/tests/libtest.c]: Just nits.

**Comment by akoeplinger** [comment]
@monojenkins build

**Pull request merged**